cmd/lotus-shed/indexes.go

```diff
@@ -483,87 +492,145 @@ var inspectEventsCmd = &cli.Command{
 		if err != nil {
 			return err
 		}
+		stmtSelectEventIdAndEmitter, err := db.Prepare(selectEventIdAndEmitter)
+		if err != nil {
+			return err
+		}
+		stmtSelectEventEntries, err := db.Prepare(selectEventEntries)
+		if err != nil {
+			return err
+		}
 
-		processHeight := func(ctx context.Context, ts *types.TipSet, receipts []*types.MessageReceipt) error {
-			tsKeyCid, err := ts.Key().Cid()
+		processHeight := func(ctx context.Context, messages []lapi.Message, receipts []*types.MessageReceipt) error {
+			tsKeyCid, err := currTs.Key().Cid()
 			if err != nil {
-				return fmt.Errorf("failed to get tipset key cid: %w", err)
+				return xerrors.Errorf("failed to get tipset key cid: %w", err)
 			}
 
-			var expectEvents int
-			var expectEntries int
+			var problems []string
+			var hasEvents bool
 
-			for _, receipt := range receipts {
-				if receipt.ExitCode != exitcode.Ok || receipt.EventsRoot == nil {
-					continue
+			checkEventAndEntryCounts := func() error {
+				// compare by counting events, using ChainGetEvents to load the events from the chain
+				expectEvents, expectEntries, err := chainEventAndEntryCountsAt(ctx, currTs, receipts, api)
+				if err != nil {
+					return err
 				}
-				events, err := api.ChainGetEvents(ctx, *receipt.EventsRoot)
+				if expectEvents > 0 {
+					hasEvents = true
```

Comment:
`hasEvents` is set in multiple places but it's only real use is in logging here:

```
if err := stmtTipsetSeen.QueryRow(tsKeyCid.Bytes()).Scan(&seenHeight, &seenReverted); err != nil {
				if err == sql.ErrNoRows {
					if hasEvents {
						problems = append(problems, "not in events_seen table")
					} else {
						problems = append(problems, "zero-event epoch not in events_seen table")
					}
```

Can we simplify how `hasEvents` is set -> set it only once and preferably during initialisation.

Reply:
I've removed one of the places where it's set, so it's only set in one place now (that was a remnant from a previous refactor that allowed for this code to go either counting or amt checking, but now it's just amt checking with counting if we need it). But we still need it to be set the first time we encounter a message with >0 events and we don't know that ahead of time. The only other way I can think to do that would be to check all of the `EventsRoot` values for each receipt and see if they are either `nil` or point to the 0-length AMT (I'm actually not sure of the fvm behaviour here - does it set that to `nil` when there are no events or set it to the CID of the 0-length AMT?). But there's even a loop in there, so I may as well do the `has` return on the AMT builder and use that.

Reply:
I see. Yeah makes sense.